context/sm_context.go

```diff
@@ -681,3 +686,70 @@ func (smContext *SMContext) CommitSmPolicyDecision(status bool) error {
 	//TODO
 	return nil
 }
+
+func (smContext *SMContext) getSmCtxtUpf() (name, ip string) {
+	var upfName, upfIP string
+	if smContext.SMContextState == SmStateActive {
+
+		if smContext.Tunnel != nil {
+			//Set UPF FQDN name if provided else IP-address
+			if smContext.Tunnel.DataPathPool[1].FirstDPNode.UPF.NodeID.NodeIdType == pfcpType.NodeIdTypeFqdn {
+				upfName = string(smContext.Tunnel.DataPathPool[1].FirstDPNode.UPF.NodeID.NodeIdValue)
+				upfName = strings.Split(upfName, ".")[0]
+				upfIP = smContext.Tunnel.DataPathPool[1].FirstDPNode.UPF.NodeID.ResolveNodeIdToIp().String()
+			} else {
+				upfName = smContext.Tunnel.DataPathPool[1].FirstDPNode.UPF.GetUPFIP()
+				upfIP = smContext.Tunnel.DataPathPool[1].FirstDPNode.UPF.GetUPFIP()
+			}
+		}
+	}
+	return upfName, upfIP
+}
+
+//Collect Ctxt info and publish on Kafka stream
+func (smContext *SMContext) PublishSmCtxtInfo() {
+	var op mi.SubscriberOp
+	kafkaSmCtxt := mi.CoreSubscriber{}
+
+	//Populate kafka sm ctxt struct
+	kafkaSmCtxt.Imsi = smContext.Supi
+	kafkaSmCtxt.IPAddress = smContext.PDUAddress.Ip.String()
+	kafkaSmCtxt.SmfSubState, op = mapPduSessStateToMetricStateAndOp(smContext.SMContextState)
+	kafkaSmCtxt.SmfId = smContext.Ref
+	kafkaSmCtxt.Slice = "sd:" + smContext.Snssai.Sd + " sst:" + strconv.Itoa(int(smContext.Snssai.Sst))
+	kafkaSmCtxt.Dnn = smContext.Dnn
+	kafkaSmCtxt.UpfName, kafkaSmCtxt.UpfAddr = smContext.getSmCtxtUpf()
+	//kafkaSmCtxt.SmfId = SMF_Self().NfInstanceID
```

Comment:
Do you want to get rid of commented code ?

Reply:
Yes, done